File: firebase/datafix/fix_missing_profile_fields.js
// Firebase Admin SDK script to fix missing profile fields
const admin = require('firebase-admin');
const serviceAccount = require('../serviceAccountKey.json');

// Initialize Firebase
admin.initializeApp({
  credential: admin.credential.cert(serviceAccount)
});

const firestore = admin.firestore();

async function fixMissingProfileFields() {
  console.log('🔧 Starting profile field fix process...');
  
  // Track stats
  let totalProfiles = 0;
  let profilesUpdated = 0;
  const missingFieldCounts = {};
  
  // Default values for missing fields
  const defaultValues = {
    displayName: 'HIVE User',
    username: '',  // Special handling for username below
    profileImageUrl: 'https://firebasestorage.googleapis.com/v0/b/hive-social-app.appspot.com/o/default_profile.png',
    bio: '',
    location: '',
    website: '',
    interests: [],
    following: [],
    followers: [],
    createdAt: admin.firestore.Timestamp.now(),
    notificationSettings: {
      events: true,
      messages: true,
      follows: true,
      mentions: true
    }
  };
  
  try {
    // Get all users
    const usersSnapshot = await firestore.collection('users').get();
    totalProfiles = usersSnapshot.size;
    
    console.log(`📊 Processing ${totalProfiles} user profiles...`);
    
    // Process each user document
    for (const userDoc of usersSnapshot.docs) {
      const userId = userDoc.id;
      const userData = userDoc.data();
      
      console.log(`👤 Checking user: ${userId}`);
      
      let needsUpdate = false;
      const updates = {};
      
      // Special handling for username - use userId as default if missing
      if (!userData.username || userData.username === '') {
        updates.username = `user_${userId}`.substring(0, 15);  // Ensure username isn't too long
        missingFieldCounts['username'] = (missingFieldCounts['username'] || 0) + 1;
        needsUpdate = true;
        console.log(`  - Adding missing username: ${updates.username}`);
      }
      
      // Check and add all other missing fields
      Object.entries(defaultValues).forEach(([field, defaultValue]) => {
        if (field !== 'username' && (!userData[field] || userData[field] === null)) {
          updates[field] = defaultValue;
          missingFieldCounts[field] = (missingFieldCounts[field] || 0) + 1;
          needsUpdate = true;
          console.log(`  - Adding missing field: ${field}`);
        }
      });
      
      // Update document if needed
      if (needsUpdate) {
        try {
          await firestore.collection('users').doc(userId).update(updates);
          profilesUpdated++;
          console.log(`  ✅ Updated profile for user: ${userId}`);
        } catch (e) {
          console.log(`  ❌ Error updating profile for user ${userId}: ${e}`);
        }
      } else {
        console.log(`  ✓ Profile has all required fields`);
      }
    }
    
    // Check for duplicate usernames and fix if found
    console.log('\n🔍 Checking for duplicate usernames...');
    const usernameToUserIds = {};
    
    // Get all users again after updates
    const updatedUsersSnapshot = await firestore.collection('users').get();
    
    // Build username to userId mapping
    for (const userDoc of updatedUsersSnapshot.docs) {
      const userId = userDoc.id;
      const userData = userDoc.data();
      
      if (userData.username) {
        const username = userData.username;
        
        usernameToUserIds[username] = usernameToUserIds[username] || [];
        usernameToUserIds[username].push(userId);
      }
    }
    
    // Fix duplicate usernames
    let duplicateUsernamesFixed = 0;
    
    for (const [username, userIds] of Object.entries(usernameToUserIds)) {
      if (userIds.length > 1) {
        console.log(`  ⚠️ Found duplicate username: ${username} used by ${userIds.length} users`);
        
        // Update all but the first user with the duplicate username
        for (let i = 1; i < userIds.length; i++) {
          const userId = userIds[i];
          const newUsername = `${username}_${i}`;
          
          try {
            await firestore.collection('users').doc(userId).update({
              username: newUsername
            });
            
            duplicateUsernamesFixed++;
            console.log(`  ✅ Changed username for user ${userId} from ${username} to ${newUsername}`);
          } catch (e) {
            console.log(`  ❌ Error updating username for user ${userId}: ${e}`);
          }
        }
      }
    }
    
    // Print summary
    console.log('\n📊 Summary:');
    console.log(`  - Total profiles processed: ${totalProfiles}`);
    console.log(`  - Profiles updated: ${profilesUpdated}`);
    console.log(`  - Duplicate usernames fixed: ${duplicateUsernamesFixed}`);
    console.log('\n  Missing field counts:');
    Object.entries(missingFieldCounts).forEach(([field, count]) => {
      console.log(`    - ${field}: ${count}`);
    });
    
  } catch (e) {
    console.log(`❌ Error fixing missing profile fields: ${e}`);
  } finally {
    // End the Firebase app
    admin.app().delete();
  }
}

// Run the function
fixMissingProfileFields(); 

File: firebase/datafix/fix_missing_profile_fields.dart
import 'package:firebase_core/firebase_core.dart';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/material.dart';

/// This script ensures all user profiles have the required fields.
/// It adds missing fields with default values to maintain data consistency
/// and prevent app crashes due to missing fields.
/// 
/// Required fields include:
/// - displayName: String
/// - username: String
/// - profileImageUrl: String
/// - bio: String
/// - interests: List<String>
/// - joinDate: Timestamp
/// - lastActive: Timestamp
/// - connections: Map
/// - badges: List
/// - settings: Map
void main() async {
  // Initialize Flutter for Firebase
  WidgetsFlutterBinding.ensureInitialized();

  // Initialize Firebase
  await Firebase.initializeApp();
  
  print('🔧 Starting profile field fix process...');
  await fixMissingProfileFields();
  print('✅ Profile field fix process completed!');
}

/// Fixes missing fields in user profiles
Future<void> fixMissingProfileFields() async {
  final FirebaseFirestore firestore = FirebaseFirestore.instance;
  
  try {
    // Get all users
    print('📋 Fetching all user profiles from Firestore...');
    final QuerySnapshot usersSnapshot = await firestore.collection('users').get();
    print('🔍 Found ${usersSnapshot.docs.length} user profiles to process');
    
    // Track statistics
    int totalProfiles = usersSnapshot.docs.length;
    int profilesUpdated = 0;
    Map<String, int> missingFieldCounts = {
      'displayName': 0,
      'username': 0,
      'profileImageUrl': 0,
      'bio': 0,
      'interests': 0,
      'joinDate': 0,
      'lastActive': 0,
      'connections': 0,
      'badges': 0,
      'settings': 0,
    };
    
    // Define default values for missing fields
    final Map<String, dynamic> defaultValues = {
      'displayName': 'HIVE User',
      'username': '',  // Will be handled specially
      'profileImageUrl': 'https://firebasestorage.googleapis.com/v0/b/hive-app.appspot.com/o/default_avatar.png',
      'bio': 'No bio yet',
      'interests': <String>[],
      'joinDate': FieldValue.serverTimestamp(),
      'lastActive': FieldValue.serverTimestamp(),
      'connections': {
        'following': <String>[],
        'followers': <String>[],
        'friends': <String>[],
      },
      'badges': <Map<String, dynamic>>[],
      'settings': {
        'notifications': true,
        'privateProfile': false,
        'theme': 'dark',
        'language': 'en',
      },
    };
    
    // Process each user profile
    int profileIndex = 0;
    for (final DocumentSnapshot userDoc in usersSnapshot.docs) {
      profileIndex++;
      final String userId = userDoc.id;
      final Map<String, dynamic> userData = userDoc.data() as Map<String, dynamic>;
      
      print('🔍 Processing profile $profileIndex/$totalProfiles: User $userId');
      
      // Check for missing fields
      Map<String, dynamic> updates = {};
      bool needsUpdate = false;
      
      // Special handling for username - use userId as default if missing
      if (!userData.containsKey('username') || userData['username'] == null || userData['username'] == '') {
        updates['username'] = 'user_$userId'.substring(0, 15);  // Ensure username isn't too long
        missingFieldCounts['username'] = (missingFieldCounts['username'] ?? 0) + 1;
        needsUpdate = true;
        print('  - Adding missing username: ${updates['username']}');
      }
      
      // Check and add all other missing fields
      defaultValues.forEach((field, defaultValue) {
        if (field != 'username' && (!userData.containsKey(field) || userData[field] == null)) {
          updates[field] = defaultValue;
          missingFieldCounts[field] = (missingFieldCounts[field] ?? 0) + 1;
          needsUpdate = true;
          print('  - Adding missing field: $field');
        }
      });
      
      // Update profile if needed
      if (needsUpdate) {
        try {
          await firestore.collection('users').doc(userId).update(updates);
          profilesUpdated++;
          print('  ✅ Updated profile for user $userId');
        } catch (e) {
          print('  ❌ Error updating profile for user $userId: $e');
        }
      } else {
        print('  ✓ Profile has all required fields');
      }
    }
    
    // Check for duplicate usernames and fix if found
    print('\n🔍 Checking for duplicate usernames...');
    Map<String, List<String>> usernameToUserIds = {};
    
    // Get all users again after updates
    final QuerySnapshot updatedUsersSnapshot = await firestore.collection('users').get();
    
    // Build username to userId mapping
    for (final DocumentSnapshot userDoc in updatedUsersSnapshot.docs) {
      final String userId = userDoc.id;
      final Map<String, dynamic> userData = userDoc.data() as Map<String, dynamic>;
      
      if (userData.containsKey('username') && userData['username'] != null) {
        final String username = userData['username'].toString();
        
        usernameToUserIds.putIfAbsent(username, () => []);
        usernameToUserIds[username]!.add(userId);
      }
    }
    
    // Fix duplicate usernames
    int duplicateUsernamesFixed = 0;
    
    for (final MapEntry<String, List<String>> entry in usernameToUserIds.entries) {
      final String username = entry.key;
      final List<String> userIds = entry.value;
      
      if (userIds.length > 1) {
        print('  ⚠️ Found duplicate username: $username used by ${userIds.length} users');
        
        // Update all but the first user with the duplicate username
        for (int i = 1; i < userIds.length; i++) {
          final String userId = userIds[i];
          final String newUsername = '${username}_$i';
          
          try {
            await firestore.collection('users').doc(userId).update({
              'username': newUsername
            });
            
            duplicateUsernamesFixed++;
            print('  ✅ Changed username for user $userId from $username to $newUsername');
          } catch (e) {
            print('  ❌ Error updating username for user $userId: $e');
          }
        }
      }
    }
    
    // Print summary
    print('\n📊 Summary:');
    print('  - Total profiles processed: $totalProfiles');
    print('  - Profiles updated: $profilesUpdated');
    print('  - Duplicate usernames fixed: $duplicateUsernamesFixed');
    print('\n  Missing field counts:');
    missingFieldCounts.forEach((field, count) {
      print('    - $field: $count');
    });
    
  } catch (e) {
    print('❌ Error fixing missing profile fields: $e');
  }
} 